conans/client/rest/client_routes.py

```diff
@@ -1,30 +1,49 @@
-from six.moves.urllib.parse import urlencode
+from six.moves.urllib.parse import quote, urlencode
 
 from conans.model.ref import ConanFileReference
 from conans.model.rest_routes import RestRoutes
-from conans.paths import CONAN_MANIFEST, CONANINFO
+from conans.paths import CONAN_MANIFEST, CONANINFO, ARTIFACTS_PROPERTIES_PUT_PREFIX
 
 
-def _format_ref(url, ref):
+def _format_ref(url, ref, matrix_params=""):
     url = url.format(name=ref.name, version=ref.version, username=ref.user or "_",
-                     channel=ref.channel or "_", revision=ref.revision)
+                     channel=ref.channel or "_", revision=ref.revision,
+                     matrix_params=matrix_params)
     return url
 
 
-def _format_pref(url, pref):
+def _format_pref(url, pref, matrix_params=""):
     ref = pref.ref
     url = url.format(name=ref.name, version=ref.version, username=ref.user or "_",
                      channel=ref.channel or "_", revision=ref.revision, package_id=pref.id,
-                     p_revision=pref.revision)
+                     p_revision=pref.revision, matrix_params=matrix_params)
     return url
 
 
-routes = RestRoutes()
+def _remove_put_prefix(artifacts_properties):
+    len_prefix = len(ARTIFACTS_PROPERTIES_PUT_PREFIX)
+    for key, value in artifacts_properties.items():
+        if key.startswith(ARTIFACTS_PROPERTIES_PUT_PREFIX):
+            key = key[len_prefix:]
+        yield key, value
+
+
+routes = RestRoutes(matrix_params=True)
 
 
 class ClientCommonRouter(object):
     """Search urls shared between v1 and v2"""
 
+    def __init__(self, base_url, artifacts_properties):
+        self.base_url = base_url
+
+        if artifacts_properties:
+            matrix_params = dict(_remove_put_prefix(artifacts_properties))
```

Comment:
What would happen if we send both?

Reply:
Artifactory adds all the `properties` in this order: (1) headers, (2) files and (3) matrix params. I think, but I need to check (I'll add it to a TODO list), that duplicated items are not stored/shown.